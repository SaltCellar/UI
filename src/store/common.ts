import { MapDispatchToProps, MapStateToProps } from 'react-redux';
import { Action } from 'redux';
import { ThunkAction as ReduxThunkAction } from 'redux-thunk';
import { RootState } from './rootReducer';

export const enum FetchStatus {
  'none',
  'inProgress',
  'complete',
}

export function createMapStateToProps<OwnProps, StateProps>(
  mapStateToProps: MapStateToProps<StateProps, OwnProps, RootState>
) {
  return mapStateToProps;
}

export function createMapDispatchToProps<OwnProps, DispatchProps>(
  mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps>
) {
  return mapDispatchToProps;
}

export type ThunkAction<A extends Action = any, R = any> = ReduxThunkAction<
  R,
  RootState,
  void,
  A
>;
