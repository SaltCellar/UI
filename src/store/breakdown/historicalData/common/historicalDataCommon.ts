import { ReportPathsType, ReportType } from 'api/reports/report';

// eslint-disable-next-line no-shadow
export const enum HistoricalDataWidgetType {
  cost = 'cost', // This type displays historical cost chart
  trend = 'trend', // This type displays historical trend chart
  usage = 'usage', // This type displays historical usage chart
}

export interface HistoricalDataWidget {
  id: number;
  reportPathsType: ReportPathsType; // Report URL path
  reportType: ReportType; // Report type; cost, storage, etc.
  type: HistoricalDataWidgetType;
}
