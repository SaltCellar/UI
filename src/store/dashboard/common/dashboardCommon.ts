import { ReportType } from 'api/reports/report';

export const enum DashboardChartType {
  cost = 'cost',
  trend = 'trend',
  usage = 'usage',
}

export const enum DashboardPerspective {
  aws = 'aws',
  awsCloud = 'aws_cloud', // Aws filtered by Ocp
  azure = 'azure',
  azureCloud = 'azure_cloud', // Azure filtered by Ocp
  ocp = 'ocp',
  ocpCloud = 'ocp_cloud', // All filtered by Ocp
  ocpUsage = 'ocp_usage',
  ocpSupplementary = 'ocp_supplementary',
}

export interface ValueFormatOptions {
  fractionDigits?: number;
}

export interface DashboardWidget<T> {
  availableTabs?: T[];
  chartType?: DashboardChartType;
  currentTab?: T;
  details: {
    costKey?: string /** i18n key */;
    formatOptions: ValueFormatOptions;
    labelKey?: string /** i18n key */;
    requestFormatOptions?: {
      fractionDigits?: number;
    };
    requestKey?: string;
    showUnits?: boolean;
    showUsageLegendLabel?: boolean;
    units?: string;
    usageFormatOptions?: ValueFormatOptions;
    usageKey?: string /** i18n key */;
  };
  filter?: {
    limit?: number;
    service?: string;
    service_name?: string;
  };
  id: number;
  isDetailsLink?: boolean;
  isHorizontal?: boolean;
  isUsageFirst?: boolean;
  perspective?: DashboardPerspective;
  reportType: ReportType;
  /** i18n key for the title. passed { startDate, endDate, month, time } */
  titleKey: string;
  tabsFilter?: {
    limit?: number;
    service?: string;
    service_name?: string;
  };
  trend: {
    comparison: string;
    titleKey: string;
    type: number;
    formatOptions: ValueFormatOptions;
  };
  topItems?: {
    formatOptions: {};
  };
}
