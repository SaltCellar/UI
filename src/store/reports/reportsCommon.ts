import { ReportType } from 'api/reports';

export const reportsStateKey = 'reports';

export function getReportId(type: ReportType, query: string) {
  return `${type}--${query}`;
}
