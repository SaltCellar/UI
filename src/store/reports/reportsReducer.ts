import { Report } from 'api/reports';
import { AxiosError } from 'axios';
import { ActionType, getType } from 'typesafe-actions';
import { FetchStatus } from '../common';
import {
  fetchReportFailure,
  fetchReportRequest,
  fetchReportSuccess,
} from './reportsActions';

export interface CachedReport extends Report {
  timeRequested: number;
}

export type ReportsState = Readonly<{
  byId: Map<string, CachedReport>;
  fetchStatus: Map<string, FetchStatus>;
  errors: Map<string, AxiosError>;
}>;

const defaultState: ReportsState = {
  byId: new Map(),
  fetchStatus: new Map(),
  errors: new Map(),
};

export type ReportsAction = ActionType<
  | typeof fetchReportFailure
  | typeof fetchReportRequest
  | typeof fetchReportSuccess
>;

export function reportsReducer(
  state = defaultState,
  action: ReportsAction
): ReportsState {
  switch (action.type) {
    case getType(fetchReportRequest):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(
          action.payload.reportId,
          FetchStatus.inProgress
        ),
      };

    case getType(fetchReportSuccess):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(
          action.meta.reportId,
          FetchStatus.complete
        ),
        byId: new Map(state.byId).set(action.meta.reportId, {
          ...action.payload,
          timeRequested: Date.now(),
        }),
        errors: new Map(state.errors).set(action.meta.reportId, null),
      };

    case getType(fetchReportFailure):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(
          action.meta.reportId,
          FetchStatus.complete
        ),
        errors: new Map(state.errors).set(action.meta.reportId, action.payload),
      };
    default:
      return state;
  }
}
