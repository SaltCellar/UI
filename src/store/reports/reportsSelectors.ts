import { ReportType } from 'api/reports';
import { RootState } from '../rootReducer';
import { getReportId, reportsStateKey } from './reportsCommon';

export const selectReportsState = (state: RootState) => state[reportsStateKey];

export const selectReport = (
  state: RootState,
  reportType: ReportType,
  query: string
) => selectReportsState(state).byId.get(getReportId(reportType, query));

export const selectReportFetchStatus = (
  state: RootState,
  reportType: ReportType,
  query: string
) => selectReportsState(state).fetchStatus.get(getReportId(reportType, query));

export const selectReportError = (
  state: RootState,
  reportType: ReportType,
  query: string
) => selectReportsState(state).errors.get(getReportId(reportType, query));
