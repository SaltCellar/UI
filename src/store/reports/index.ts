import * as reportsActions from './reportsActions';
import { reportsStateKey } from './reportsCommon';
import {
  CachedReport,
  ReportsAction,
  reportsReducer,
  ReportsState,
} from './reportsReducer';
import * as reportsSelectors from './reportsSelectors';

export {
  ReportsAction,
  CachedReport,
  reportsActions,
  reportsReducer,
  reportsSelectors,
  ReportsState,
  reportsStateKey,
};
