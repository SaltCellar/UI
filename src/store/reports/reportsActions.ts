import { Report, ReportType, runReport } from 'api/reports';
import { AxiosError } from 'axios';
import addMinutes from 'date-fns/add_minutes';
import isBefore from 'date-fns/is_before';
import { ThunkAction } from 'redux-thunk';
import { createStandardAction } from 'typesafe-actions';
import { FetchStatus } from '../common';
import { RootState } from '../rootReducer';
import { getReportId } from './reportsCommon';
import { selectReport, selectReportFetchStatus } from './reportsSelectors';

const expirationMinutes = 30;

interface ReportActionMeta {
  reportId: string;
}

export const fetchReportRequest = createStandardAction('reports/request')<
  ReportActionMeta
>();
export const fetchReportSuccess = createStandardAction('reports/success')<
  Report,
  ReportActionMeta
>();
export const fetchReportFailure = createStandardAction('reports/failure')<
  AxiosError,
  ReportActionMeta
>();

export function fetchReport(
  reportType: ReportType,
  query: string
): ThunkAction<void, RootState, void, any> {
  return (dispatch, getState) => {
    if (!isReportExpired(getState(), reportType, query)) {
      return;
    }

    const meta: ReportActionMeta = {
      reportId: getReportId(reportType, query),
    };

    dispatch(fetchReportRequest(meta));
    runReport(reportType, query)
      .then(res => {
        dispatch(fetchReportSuccess(res.data, meta));
      })
      .catch(err => {
        dispatch(fetchReportFailure(err, meta));
      });
  };
}

function isReportExpired(
  state: RootState,
  reportType: ReportType,
  query: string
) {
  const report = selectReport(state, reportType, query);
  const fetchStatus = selectReportFetchStatus(state, reportType, query);

  if (fetchStatus === FetchStatus.inProgress) {
    return false;
  }

  if (!report) {
    return true;
  }

  return isBefore(
    new Date(),
    addMinutes(report.dateRequested, expirationMinutes)
  );
}
