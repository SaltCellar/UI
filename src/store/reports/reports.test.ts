jest.mock('api/reports');

import { Report, ReportType, runReport } from 'api/reports';
import { wait } from 'testUtils';
import { FetchStatus } from '../common';
import { createMockStoreCreator } from '../mockStore';
import * as actions from './reportsActions';
import { reportsStateKey } from './reportsCommon';
import { reportsReducer } from './reportsReducer';
import * as selectors from './reportsSelectors';

const createReportsStore = createMockStoreCreator({
  [reportsStateKey]: reportsReducer,
});

const runReportMock = runReport as jest.Mock;

const mockReport: Report = {
  data: [],
  total: {
    value: 100,
    units: 'USD',
  },
};

const query = 'query';
const reportType = ReportType.cost;

runReportMock.mockResolvedValue({ data: mockReport });
global.Date.now = jest.fn(() => 12345);

test('default state', () => {
  const store = createReportsStore();
  expect(selectors.selectReportsState(store.getState())).toMatchSnapshot();
});

test('fetch report success', async () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportType, query));
  expect(runReportMock).toBeCalled();
  expect(
    selectors.selectReportFetchStatus(store.getState(), reportType, query)
  ).toBe(FetchStatus.inProgress);
  await wait();
  const finishedState = store.getState();
  expect(
    selectors.selectReport(finishedState, reportType, query)
  ).toMatchSnapshot();
  expect(
    selectors.selectReportFetchStatus(finishedState, reportType, query)
  ).toBe(FetchStatus.complete);
  expect(selectors.selectReportError(finishedState, reportType, query)).toBe(
    null
  );
});

test('fetch report failure', async () => {
  const store = createReportsStore();
  const error = Symbol('report error');
  runReportMock.mockRejectedValueOnce(error);
  store.dispatch(actions.fetchReport(reportType, query));
  expect(runReport).toBeCalled();
  expect(
    selectors.selectReportFetchStatus(store.getState(), reportType, query)
  ).toBe(FetchStatus.inProgress);
  await wait();
  const finishedState = store.getState();
  expect(
    selectors.selectReportFetchStatus(finishedState, reportType, query)
  ).toBe(FetchStatus.complete);
  expect(selectors.selectReportError(finishedState, reportType, query)).toBe(
    error
  );
});

test('does not fetch report if the request is in progress', () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportType, query));
  store.dispatch(actions.fetchReport(reportType, query));
  expect(runReport).toHaveBeenCalledTimes(1);
});

test('report is not refetched if it has not expired', async () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportType, query));
  await wait();
  store.dispatch(actions.fetchReport(reportType, query));
  expect(runReport).toHaveBeenCalledTimes(1);
});
