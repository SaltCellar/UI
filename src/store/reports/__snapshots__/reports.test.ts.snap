// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
Object {
  "byId": Map {},
  "errors": Map {},
  "fetchStatus": Map {},
}
`;

exports[`fetch report success 1`] = `
Object {
  "data": Array [],
  "dateRequested": 2018-01-01T00:00:00.000Z,
  "total": Object {
    "units": "USD",
    "value": 100,
  },
}
`;
